### ISSUE TITLE
View consistency for PrimTorch+nvFuser tests

### ISSUE THREAD

**Issue by IvanYashchuk** [pytorch/pytorch#84863]
### 🚀 The feature, motivation and pitch

`test/test_ops.py:_ref_test_helper` checks whether the output of a reference implementation returns the same `._is_view()` as eager.

The nvFuser execution function always returns a fresh copy currently so `tensor._is_view()` is False. These tests are currently skipped.
How important is view consistency? It's not yet clear whether it's feasible to implement for nvFuser execution and how difficult it is.

### Additional context

Latest master.

cc @ezyang @mruberry @ngimel @Lezcano @fdrocha

**Comment by Lezcano** [comment]
What about setting `validate_view_consistency=False` for all nvFuser ops?

**Comment by ezyang** [comment]
So, this is not really an edge case that nvfuser optimizes, but if you have a subgraph where the output is a view of the input, you are obligated to actually return a view, because a user may mutate through the view later and you have to maintain the aliasing relationship.

**Comment by ngimel** [comment]
Should functionalization get rid of those edge cases and replace them with explicit copies to input to maintain aliasing?

**Comment by ezyang** [comment]
I suppose yes, for outputs which are views of inputs, it would be good if these "moved" out of the main graph into the epilogue (which is also doing the explicit copies for mutated inputs). @bdhirsh, IIUC, this doesn't happen yet right?

**Comment by bdhirsh** [comment]
> Should functionalization get rid of those edge cases and replace them with explicit copies to input to maintain aliasing?

Yep - for "normal" graphs, if you've run functionalization first before passing your graph (or subgraphs) off to nvfuser, then there shouldn't be any mutations in the graph. So if you had a subgraph like `x.view(....).view(...)` that nvfuser ended up re-writing to return a copy, you wouldn't have to worry about mutations causing problems.

@ezyang maybe I'm missing something - you're right that AOTAutograd currently does not yet do the thing where `copy_()` calls are moved out of the main graph into an epilogue. But I'm not sure that's actually a problem in the context of this issue? Say I have a program that mutates its inputs, that contains a subgraph that aliases outputs, which nvfuser might want to muck with:
```
def program(x):
    def subgraph(tmp):
        return tmp.t().t()
    # basic (probably unrealistic) example of a "subgraph" whose output aliases its input
    y = subgraph(tmp)
    # uh-oh; what if nvfuser re-writes the subgraph so it's no longer a view of the input?
    y.add_(1)
    return y
```

I think this would still be fine, since with functionalization you would end up with:
```
def program(x):
    y = x.t().t()
    y_updated = y.add(1)
    x_updated = y_updated.t().t()
    # At this point, it doesn't matter if nvfuser mucks with `y_updated.t().t()` to no longer return a copy.
    # the functionalized graph doesn't rely on that aliasing behavior anymore, since it directly
    # copies `x_updated` into the input.
    x.copy_(x_updated)
    return y
```
```

**Comment by ezyang** [comment]
Nah, this is not what I'm worried about. It's simpler.

```
def program(x):
  y = x.view(10)
  z = y * 2
  return y, z
```

suppose nvfuser decomposes view into view_copy

```
def program(x):
  y = x.view_copy(10)
  z = y * 2
  return y, z
```

this program's returned y no longer aliases with x. So this is wrong. But if you detect outputs that alias inputs and manually compute them before returning, it doesn't matter if the compiler goops up aliasing internally (which it should be in its rights to do, because the point of functionalization is to make it so you don't have to care about aliasing.)

**Comment by IvanYashchuk** [comment]
Where is this functionalization implemented? `functionalize` from Functorch doesn't do anything with view.
```py
In [1]: import torch

In [2]: from torch.fx.experimental.proxy_tensor import make_fx

In [3]: from functorch.experimental import functionalize

In [4]: def program(x: torch.Tensor):
   ...:   y = x.view(10)
   ...:   z = y * 2
   ...:   return y, z
   ...:

In [5]: x = torch.randn(2, 5, device="cuda")

In [6]: gm = make_fx(functionalize(program))(x)

In [7]: gm.graph.print_tabular()
opcode         name    target             args            kwargs
-------------  ------  -----------------  --------------  --------
placeholder    x_1     x_1                ()              {}
call_function  view    aten.view.default  (x_1, [10])     {}
call_function  mul     aten.mul.Tensor    (view, 2)       {}
output         output  output             ((view, mul),)  {}
```

If I want to alias output to input is there a way to do so programmatically from Python? I can create tensors that would share the storage but would return False for `tensor._is_view()`.

**Comment by bdhirsh** [comment]
> But if you detect outputs that alias inputs and manually compute them before returning, it doesn't matter if the compiler goops up aliasing internally (which it should be in its rights to do, because the point of functionalization is to make it so you don't have to care about aliasing.)

ah ok. This sounds reasonable - and I guess you'd also need to shunt those "manually recomputed views" into an epilogue that way nvfuser can't muck with it, like you said. I don't think there's a github issue yet for stashing extra views / copies into an epilogue in AOTAutograd, so I'll make one.

Although two things:

(1) Functionalization doesn't really know about tracing, graphs, or "stashing stuff in epilogues" - it's just a function -> function transform. (As with my [prototype](https://github.com/pytorch/pytorch/pull/82602) PR), I think AOTAutograd needs to directly handle the logic around creating / stashing stuff in an epilogue. Maybe AOTAutograd should also be in charge of "detecting outputs that alias inputs, manually computing them and stashing them in the epilogue"? We could update functionalization to directly emit extra `as_strided()` calls when it sees outputs that alias inputs, but then the logic that creates the epilogue is on the hook for finding those calls in the graph (instead of generating them itself).

On the other hand, functionalization also has a mode `remove="mutations_and_views"` (not used by dynamo, but used by LTC) that also replaces views with view copies. I think you're right that the logic would need to live directly in functionalization when that flag is set, or else functionalization will literally be wrong. Although this seems lower priority, since `remove="mutations_and_views"` is currently only really intended to be used by LTC and XLA - and it actually isn't a problem for them, because you aren't allowed to have views of tensors across multiple devices.


(2) I'm mostly just curious on why nvfuser sometimes replaces views with view copies (since presumably that hurts perf, if you're making a copy). Maybe sometimes when the input is strided in a weird way, it would rather create a contiguous copy?

> Where is this functionalization implemented?

The stuff above isn't currently implemented for functionalization (but it probably should be!)

**Comment by ezyang** [comment]
Yeah I'm less concerned about where we put the logic; AOTAutograd would be fine too

**Comment by bdhirsh** [comment]
Issue here: https://github.com/pytorch/pytorch/issues/85036

**Comment by jjsjann123** [comment]
> (2) I'm mostly just curious on why nvfuser sometimes replaces views with view copies (since presumably that hurts perf, if you're making a copy). Maybe sometimes when the input is strided in a weird way, it would rather create a contiguous copy?

well, nvfuser mostly tries to inline computation. 1. If an intermediate tensor is viewed, making it a copy actually doesn't hurt our perf, since intermediates are going to be hold in register anyway. 2. For output tensor, our scheduler currently has some assumptions on tensor memory layout. We don't yet support explicit output strides at this moment. We'll need some work on our scheduler in order to generate kernel with good performance on arbitrary strides. Also, aliasing is hard and treating `view` as `view_copy` allows us to ignore that for a little bit longer.
